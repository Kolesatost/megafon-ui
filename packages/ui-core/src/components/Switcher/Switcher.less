@import (reference) '~styles/base.less';

@b: e('.mfui-switcher');

@{b} {
    top: 22px;
    right: 28px;

    width: 40px;
    height: 24px;
    border-radius: 15.5px;

    background-color: var(--spbSky1);
    cursor: pointer;

    transition: all 0.3s;

    &_no-touch:hover {
        background-color: var(--spbSky1);
    }

    &__pointer {
        position: relative;
        top: 2px;
        left: 2px;

        width: 20px;
        height: 20px;
        border-radius: 50%;

        background-color: var(--base);

        transition: all 0.3s;
    }

    &_checked {
        background-color: var(--brandGreen);

        @{b}__pointer {
            left: 18px;
        }
    }

    &_checked&_no-touch:hover {
        background-color: var(--buttonHoverGreen);
    }

    &_disabled {
        background-color: var(--spbSky1);
        cursor: default;
    }

    &_disabled&_checked&_no-touch&:hover {
        background-color: var(--spbSky1);
    }

    &_disabled &__pointer {
        background-color: var(--spbSky2);
    }
}
