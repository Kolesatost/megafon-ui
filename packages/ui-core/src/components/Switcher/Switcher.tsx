import * as React from 'react';
import { cnCreate, detectTouch, filterDataAttrs } from '@megafon/ui-helpers';
import * as PropTypes from 'prop-types';
import './Switcher.less';

export interface ISwitcherProps {
    /** Дополнительные data атрибуты к внутренним элементам */
    dataAttrs?: {
        root?: Record<string, string>;
    };
    /** Дополнительный класс корневого элемента */
    className?: string;
    /** Управление состоянием вкл/выкл компонента */
    checked?: boolean;
    /** Управление возможностью взаимодействия с компонентом */
    disabled?: boolean;
    /** Обработчик изменения элемента */
    onChange?: (e: React.MouseEvent<HTMLDivElement>) => void;
}

const cn = cnCreate('mfui-switcher');
const Switcher: React.FC<ISwitcherProps> = ({ dataAttrs, className, checked = false, disabled = false, onChange }) => {
    const isTouch: boolean = detectTouch();

    const handleChange = (e: React.MouseEvent<HTMLDivElement>): void => {
        if (disabled) {
            return;
        }

        onChange?.(e);
    };

    return (
        <div
            {...filterDataAttrs(dataAttrs?.root)}
            className={cn(
                {
                    checked,
                    disabled,
                    'no-touch': !isTouch,
                },
                className,
            )}
            onClick={handleChange}
        >
            <div className={cn('pointer')} />
        </div>
    );
};

Switcher.propTypes = {
    dataAttrs: PropTypes.shape({
        root: PropTypes.objectOf(PropTypes.string.isRequired),
    }),
    className: PropTypes.string,
    checked: PropTypes.bool,
    disabled: PropTypes.bool,
    onChange: PropTypes.func,
};

export default Switcher;
