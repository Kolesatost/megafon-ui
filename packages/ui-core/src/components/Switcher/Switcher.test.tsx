import * as React from 'react';
import { shallow } from 'enzyme';
import Switcher from './Switcher';

describe('<Switcher />', () => {
    describe('layout', () => {
        it('should render switcher with default props', () => {
            const wrapper = shallow(<Switcher />);
            expect(wrapper).toMatchSnapshot();
        });

        it('should render switcher with external className and data-attributes', () => {
            const wrapper = shallow(
                <Switcher
                    className="external-class-name"
                    dataAttrs={{ root: { 'data-testid': 'root-test' }, input: { 'data-testid': 'input-test' } }}
                />,
            );
            expect(wrapper).toMatchSnapshot();
        });

        it('should render checked switcher', () => {
            const wrapper = shallow(<Switcher checked />);
            expect(wrapper).toMatchSnapshot();
        });

        it('should render disabled switcher', () => {
            const wrapper = shallow(<Switcher disabled />);
            expect(wrapper).toMatchSnapshot();
        });

        it('should render showLoader switcher with children', () => {
            const wrapper = shallow(<Switcher showLoader>children</Switcher>);
            expect(wrapper).toMatchSnapshot();
        });

        it('should render switcher with children with small size by left', () => {
            const wrapper = shallow(
                <Switcher showLoader textPosition="left" textSize="small">
                    children
                </Switcher>,
            );
            expect(wrapper).toMatchSnapshot();
        });

        it('should render switcher for mobile devices', () => {
            const onTouchStart = window.ontouchstart;
            window.ontouchstart = jest.fn();

            const wrapper = shallow(<Switcher />);
            expect(wrapper).toMatchSnapshot();

            window.ontouchstart = onTouchStart;
        });
    });

    describe('handleChange', () => {
        it('should call onChange id switcher is enabled', () => {
            const handleChange = jest.fn();
            const wrapper = shallow(<Switcher onChange={handleChange} />);

            const event = { target: {}, nativeEvent: {}, type: 'click' };
            wrapper.find('.mfui-switcher__input').simulate('click', event);
            expect(handleChange).toHaveBeenCalledWith(event);
        });

        it('shouldn`t call onChange id switcher is disabled', () => {
            const handleChange = jest.fn();
            const wrapper = shallow(<Switcher disabled onChange={handleChange} />);

            const event = { target: {}, nativeEvent: {}, type: 'click' };
            wrapper.find('.mfui-switcher__input').simulate('click', event);
            expect(handleChange).not.toHaveBeenCalled();
        });
    });
});
