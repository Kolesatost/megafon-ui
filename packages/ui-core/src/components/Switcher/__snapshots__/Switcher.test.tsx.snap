// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switcher /> layout should render checked switcher 1`] = `
<div
  className="mfui-switcher"
>
  <div
    className="mfui-switcher__input mfui-switcher__input_checked mfui-switcher__input_no-touch"
    onClick={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
  >
    <div
      className="mfui-switcher__pointer"
    />
  </div>
</div>
`;

exports[`<Switcher /> layout should render disabled switcher 1`] = `
<div
  className="mfui-switcher mfui-switcher_disabled"
>
  <div
    className="mfui-switcher__input mfui-switcher__input_disabled mfui-switcher__input_no-touch"
    onClick={[Function]}
    onKeyDown={[Function]}
  >
    <div
      className="mfui-switcher__pointer"
    />
  </div>
</div>
`;

exports[`<Switcher /> layout should render showLoader switcher with children 1`] = `
<div
  className="mfui-switcher"
>
  <div
    className="mfui-switcher__input mfui-switcher__input_loaded mfui-switcher__input_no-touch"
    onClick={[Function]}
    onKeyDown={[Function]}
  >
    <div
      className="mfui-switcher__loader"
    />
    <div
      className="mfui-switcher__pointer"
    />
  </div>
  <div
    className="mfui-switcher__content mfui-switcher__content_size_medium"
  >
    children
  </div>
</div>
`;

exports[`<Switcher /> layout should render switcher for mobile devices 1`] = `
<div
  className="mfui-switcher"
>
  <div
    className="mfui-switcher__input"
    onClick={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
  >
    <div
      className="mfui-switcher__pointer"
    />
  </div>
</div>
`;

exports[`<Switcher /> layout should render switcher with children with small size by left 1`] = `
<div
  className="mfui-switcher"
>
  <div
    className="mfui-switcher__content mfui-switcher__content_size_small mfui-switcher__content_left"
  >
    children
  </div>
  <div
    className="mfui-switcher__input mfui-switcher__input_loaded mfui-switcher__input_no-touch"
    onClick={[Function]}
    onKeyDown={[Function]}
  >
    <div
      className="mfui-switcher__loader"
    />
    <div
      className="mfui-switcher__pointer"
    />
  </div>
</div>
`;

exports[`<Switcher /> layout should render switcher with default props 1`] = `
<div
  className="mfui-switcher"
>
  <div
    className="mfui-switcher__input mfui-switcher__input_no-touch"
    onClick={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
  >
    <div
      className="mfui-switcher__pointer"
    />
  </div>
</div>
`;

exports[`<Switcher /> layout should render switcher with external className and data-attributes 1`] = `
<div
  className="mfui-switcher external-class-name"
  data-testid="root-test"
>
  <div
    className="mfui-switcher__input mfui-switcher__input_no-touch"
    onClick={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
  >
    <div
      className="mfui-switcher__pointer"
    />
  </div>
</div>
`;
